File: src/main/java/com/hf4all/planner/model/FuelStrip.java
package com.hf4all.planner.model;

/**
 * The HF4A Fuel Strip: a non-linear scale that maps Wet Mass positions
 * (1..32) to cumulative fuel-step distances on the "black line".
 *
 * <p>Each Wet Mass position {@code N} sits at a varying number of fuel
 * steps above its neighbour {@code N+1}. The intervals follow a rocket-
 * equation-shaped curve: large near the dry end, small near the wet end.
 * The values below are taken directly from the printed playmat in the
 * HF4A Core Rules (verified against the worked examples on pp. 14–16).
 *
 * <p>Two key conversions live here:
 * <ul>
 *   <li>{@link #stepsBetween} — how many fuel steps lie between two Wet
 *       Mass positions on the black line (the "burnable" distance).</li>
 *   <li>{@link #wetMassAt} — given a Dry Mass and the current chit
 *       position measured in fuel steps from Dry Mass, return the
 *       integer Wet Mass the chit sits at (rounded down to the nearest
 *       labelled mass; partial fractional progress into the next
 *       interval is exposed via {@link #stepsIntoInterval}).</li>
 * </ul>
 *
 * <p>Weight class is derived from the integer Wet Mass per HF4A rule H3c
 * (see {@link #weightClassMod}). The class snapshot at the start of each
 * movement determines the engine's net thrust for that movement.
 */
public final class FuelStrip {

    /** Smallest legal Dry Mass; rule F2a treats Dry Mass `0' as `1'. */
    public static final int MIN_DRY_MASS = 1;
    /** Maximum Dry Mass position on the strip (rule F2a). */
    public static final int MAX_DRY_MASS = 23;
    /** Maximum Wet Mass position on the strip (rule F3a). */
    public static final int MAX_WET_MASS = 32;

    /**
     * {@code INTERVALS[N]} = number of fuel steps between Wet Mass
     * position {@code N} and {@code N+1}. Index 0 unused; index 32
     * unused (no position above MAX_WET_MASS).
     */
    private static final int[] INTERVALS = new int[MAX_WET_MASS + 1];
    static {
        // Low-mass region (rocket-equation steep): big intervals.
        INTERVALS[1]  = 9;  // 1 ↔ 2
        INTERVALS[2]  = 6;  // 2 ↔ 3
        INTERVALS[3]  = 4;  // 3 ↔ 4
        INTERVALS[4]  = 3;  // 4 ↔ 5
        INTERVALS[5]  = 3;  // 5 ↔ 6
        // Mid region: 2 steps per mass.
        INTERVALS[6]  = 2;
        INTERVALS[7]  = 2;
        INTERVALS[8]  = 2;
        INTERVALS[9]  = 2;
        INTERVALS[10] = 2;
        // High region: 1 step per mass, all the way to MAX_WET_MASS.
        for (int i = 11; i <= MAX_WET_MASS - 1; i++) {
            INTERVALS[i] = 1;
        }
    }

    private FuelStrip() {}

    // -------------------------------------------------------------------------
    // Strip geometry
    // -------------------------------------------------------------------------

    /**
     * Fuel-step distance on the black line between two Wet Mass positions.
     * Order-independent: {@code stepsBetween(a,b) == stepsBetween(b,a)}.
     *
     * @throws IllegalArgumentException if either position is outside [1, 32]
     */
    public static int stepsBetween(int a, int b) {
        checkInRange(a);
        checkInRange(b);
        int lo = Math.min(a, b);
        int hi = Math.max(a, b);
        int sum = 0;
        for (int i = lo; i < hi; i++) sum += INTERVALS[i];
        return sum;
    }

    /**
     * Fuel-step distance from {@code dryMass} up to a given starting fuel-load
     * (number of water tanks). This is the initial {@code fuelStepsRemaining}
     * for a request specified as {@code (dryMass, fuel)} where {@code fuel} is
     * the count of tanks loaded (red-line steps).
     *
     * @throws IllegalArgumentException if dryMass + fuel exceeds {@link #MAX_WET_MASS}
     */
    public static int initialFuelSteps(int dryMass, int fuel) {
        if (dryMass < MIN_DRY_MASS || dryMass > MAX_DRY_MASS) {
            throw new IllegalArgumentException("dryMass out of range: " + dryMass);
        }
        if (fuel < 0) throw new IllegalArgumentException("fuel must be >= 0");
        int wet = dryMass + fuel;
        if (wet > MAX_WET_MASS) {
            throw new IllegalArgumentException(
                    "wetMass " + wet + " exceeds MAX_WET_MASS " + MAX_WET_MASS);
        }
        return stepsBetween(dryMass, wet);
    }

    /**
     * Validates a request specified directly in fuel-strip steps. Enforces
     * the strip's geometry: the Wet chit (at {@code dryStep + fuelSteps})
     * must land inside [0, 56].
     *
     * @throws IllegalArgumentException if either bound is violated
     */
    public static void validateFuelSteps(int dryMass, int fuelSteps) {
        if (dryMass < MIN_DRY_MASS || dryMass > MAX_DRY_MASS) {
            throw new IllegalArgumentException("dryMass out of range: " + dryMass);
        }
        if (fuelSteps < 0) throw new IllegalArgumentException("fuelSteps must be >= 0");
        int dryStep = stepsBetween(MIN_DRY_MASS, dryMass);
        int wetStep = dryStep + fuelSteps;
        int totalSteps = stepsBetween(MIN_DRY_MASS, MAX_WET_MASS);
        if (wetStep > totalSteps) {
            throw new IllegalArgumentException(
                    "wetStep " + wetStep + " exceeds total fuel-strip steps " + totalSteps
                  + " (dryMass=" + dryMass + ", fuelSteps=" + fuelSteps + ")");
        }
    }

    /**
     * Integer Wet Mass position the chit sits at (or just past), given the
     * Dry Mass and the chit's current fuel-step distance above Dry Mass.
     *
     * <p>If {@code fuelStepsRemaining} falls strictly inside an interval
     * (mid-burn fractional state), this returns the LOWER of the two
     * surrounding mass positions — i.e. the chit hasn't crossed the next
     * label yet. The fractional progress can be recovered with
     * {@link #stepsIntoInterval}.
     */
    public static int wetMassAt(int dryMass, int fuelStepsRemaining) {
        if (dryMass < MIN_DRY_MASS || dryMass > MAX_WET_MASS) {
            throw new IllegalArgumentException("dryMass out of range: " + dryMass);
        }
        if (fuelStepsRemaining < 0) {
            throw new IllegalArgumentException("fuelStepsRemaining must be >= 0");
        }
        int pos = dryMass;
        int walked = 0;
        while (pos < MAX_WET_MASS) {
            int interval = INTERVALS[pos];
            if (walked + interval > fuelStepsRemaining) break;
            walked += interval;
            pos++;
        }
        return pos;
    }

    /**
     * Fuel steps already walked into the interval above the current Wet Mass
     * label (0 if the chit sits exactly on a label).
     */
    public static int stepsIntoInterval(int dryMass, int fuelStepsRemaining) {
        int wet = wetMassAt(dryMass, fuelStepsRemaining);
        return fuelStepsRemaining - stepsBetween(dryMass, wet);
    }

    // -------------------------------------------------------------------------
    // Weight class
    // -------------------------------------------------------------------------

    // Class boundaries are defined in FUEL-STRIP STEP positions (HF4A H3c
     // playmat bands), not integer Wet Mass alone — the same integer mass
     // can span two classes when fractional positions cross a boundary.
     // Heaviest position still in each class:
     //
     //   step  8 (= WM 1+8/9)  → Wisp upper bound
     //   step 20 (= WM 4+1/3)  → Probe upper bound
     //   step 29 (= WM 8)      → Scout upper bound
     //   step 40 (= WM 16)     → Transport upper bound
     //   step 56 (= WM 32)     → Tug upper bound (= MAX_WET_MASS)
     //
    /** HF4A weight class modifier for a fuel-strip step position
     *  (0 = WM 1, 56 = WM 32). The chit's actual position on the strip
     *  determines its class — fractional positions are honoured. */
    public static int weightClassModForStep(int stepFromMass1) {
        if (stepFromMass1 <=  8) return +2; // Wisp      (WM ≤ 1+8/9)
        if (stepFromMass1 <= 20) return +1; // Probe     (WM ≤ 4+1/3)
        if (stepFromMass1 <= 29) return  0; // Scout     (WM ≤ 8)
        if (stepFromMass1 <= 40) return -1; // Transport (WM ≤ 16)
        return -2;                          // Tug       (WM ≤ 32)
    }

    /**
     * HF4A weight class modifier for a chit sitting at exact integer
     * Wet Mass. Equivalent to {@code weightClassModForStep(stepsBetween(1, wetMass))}.
     * Note: classes are defined on fuel-step positions, so this convenience
     * is only correct when the chit IS at the integer mass — not at a
     * fractional position past it.
     */
    public static int weightClassModForWetMass(int wetMass) {
        return weightClassModForStep(stepsBetween(MIN_DRY_MASS, wetMass));
    }

    /**
     * Weight class modifier given (dryMass, fuelStepsRemaining). The
     * chit's strip position is {@code dryStep + fuelStepsRemaining};
     * class is read off that position directly.
     */
    public static int weightClassMod(int dryMass, int fuelStepsRemaining) {
        int dryStep = stepsBetween(MIN_DRY_MASS, dryMass);
        return weightClassModForStep(dryStep + fuelStepsRemaining);
    }

    // -------------------------------------------------------------------------
    // Jettison ladder
    // -------------------------------------------------------------------------

    /**
     * Heaviest fuel-strip step position still in each class (Transport,
     * Scout, Probe, Wisp). Iteration order = lightening: jettisoning to
     * each successive entry requires more fuel dumped. Tug isn't an
     * entry — it's the implicit "current class" before any jettison.
     */
    private static final int[] CLASS_TOP_STEP = { 40, 29, 20, 8 };

    /**
     * The amounts (in fuel steps to jettison) that would each move the chit
     * to a *different* weight class than the current state. Returns at most
     * 4 entries. Empty if no jettison can change the class (e.g. already at
     * Wisp, or dryMass keeps the class fixed because target band requires
     * lower WM than dryMass itself).
     *
     * <p>Each returned amount is the MINIMUM jettison needed to just cross
     * into the target class — i.e. the chit lands on the heaviest position
     * still in the target class's band. Jettisoning more within the same
     * band is strictly Pareto-dominated.
     */
    public static int[] jettisonAmountsForClassChange(int dryMass, int fuelStepsRemaining) {
        int dryStep = stepsBetween(MIN_DRY_MASS, dryMass);
        int currentWetStep = dryStep + fuelStepsRemaining;
        int currentMod = weightClassModForStep(currentWetStep);

        int[] out = new int[CLASS_TOP_STEP.length];
        int n = 0;
        int lastMod = currentMod;
        for (int targetTopStep : CLASS_TOP_STEP) {
            // If dryStep already exceeds the target class's top, the chit
            // physically can't reach that band (Wet ≥ Dry).
            if (dryStep > targetTopStep) continue;

            // Already at or below the target band? Skip.
            if (currentWetStep <= targetTopStep) continue;

            int targetMod = weightClassModForStep(targetTopStep);
            if (targetMod == lastMod) continue;

            int jettison = currentWetStep - targetTopStep;
            out[n++] = jettison;
            lastMod = targetMod;
        }
        if (n == out.length) return out;
        int[] trimmed = new int[n];
        System.arraycopy(out, 0, trimmed, 0, n);
        return trimmed;
    }

    private static void checkInRange(int wetMass) {
        if (wetMass < MIN_DRY_MASS || wetMass > MAX_WET_MASS) {
            throw new IllegalArgumentException("wet/dry mass position out of range [1,32]: " + wetMass);
        }
    }
}

File: src/main/resources/static/js/fuelStrip.js
// HF4A fuel strip intervals: indexed by Wet Mass position N, the value is
// the number of black-line fuel steps between WM N and WM N+1. Matches
// FuelStrip.INTERVALS on the backend. Indices 0 and 32 are unused.
export const FUEL_STRIP_INTERVALS = [
    0,                          // 0 (unused)
    9, 6, 4, 3, 3,              //  1↔2 ..  5↔6
    2, 2, 2, 2, 2,              //  6↔7 .. 10↔11
    1, 1, 1, 1, 1, 1, 1, 1, 1,  // 11↔12 .. 19↔20
    1, 1, 1, 1, 1, 1, 1, 1, 1,  // 20↔21 .. 28↔29
    1, 1, 1                     // 29↔30 .. 31↔32
];

// Cumulative step count from Wet Mass 1 up to Wet Mass 32 inclusive.
// Used by the endpoint-fuel-strip chit editor mapping.
export const STRIP_MAX_STEP = 56;

// Mirrors FuelStrip.weightClassModForWetMass on the backend (HF4A H3c).
export function weightClassMod(wetMass) {
    if (wetMass <= 1) return +2;
    if (wetMass === 2) return +1;
    if (wetMass === 3) return 0;
    if (wetMass <= 5) return -1;
    return -2;
}

// HF4A weight class name for a given Wet Mass.
export function weightClassName(wetMass) {
    if (wetMass <= 1) return 'Wisp';
    if (wetMass === 2) return 'Probe';
    if (wetMass === 3) return 'Scout';
    if (wetMass <= 5) return 'Transport';
    return 'Tug';
}

// Express `fuelSteps` remaining above `dryMass` as
//   "<full wet-mass jumps> + <steps into current jump>/<size of current jump>
//    (total steps <fuelSteps>)"
// Example: 1 step above dry mass 4 → "0 + 1/3 (total steps 1)".
export function formatFuelOnStrip(dryMass, fuelSteps) {
    let remaining = fuelSteps;
    let pos = Math.max(1, dryMass | 0);
    let complete = 0;
    while (pos < 32 && remaining >= FUEL_STRIP_INTERVALS[pos]) {
        remaining -= FUEL_STRIP_INTERVALS[pos];
        complete++;
        pos++;
    }
    const size = pos < 32 ? FUEL_STRIP_INTERVALS[pos] : 1;
    return complete + ' + ' + remaining + '/' + size
            + ' (total steps ' + fuelSteps + ')';
}

// Cumulative fuel-step count from Wet Mass 1 up to (and INCLUDING) M.
export function massToStripStep(mass) {
    let s = 0;
    for (let i = 1; i < mass; i++) s += FUEL_STRIP_INTERVALS[i];
    return s;
}

// Step indices on the strip that correspond to integer Wet Mass positions
// (1..32). Useful for snapping the Dry chit during drag — Dry can ONLY
// land on integer-mass positions per HF4A F2.
export const INTEGER_MASS_STEPS = (() => {
    const a = [];
    for (let m = 1; m <= 32; m++) a.push(massToStripStep(m));
    return a;
})();

// Pretty-print a fuel-step distance as "N" or "M+num/den" relative to a
// given dryMass. The denominator is the size of the interval at the wet
// integer mass — see FUEL_STRIP_INTERVALS. Examples:
//   formatFuelSteps(0, 4)   → "0"
//   formatFuelSteps(3, 4)   → "1"        (3 = stepsBetween(4,5))
//   formatFuelSteps(4, 4)   → "1+1/3"    (4 steps from dry=4 → wet between 5..6)
//   formatFuelSteps(56, 1)  → "31"       (full strip)
export function formatFuelSteps(fuelSteps, dryMass) {
    if (fuelSteps <= 0) return '0';
    let remaining = fuelSteps;
    let pos = Math.max(1, dryMass | 0);
    let intPart = 0;
    while (pos < 32 && remaining >= FUEL_STRIP_INTERVALS[pos]) {
        remaining -= FUEL_STRIP_INTERVALS[pos];
        intPart++;
        pos++;
    }
    if (remaining === 0) return String(intPart);
    const den = FUEL_STRIP_INTERVALS[pos];
    return intPart + '+' + remaining + '/' + den;
}

// Parse a fuel-input string ("5", "1+5/6", "+5/6", "0+3/4", with optional
// whitespace) into a fuel-step count given the dryMass context.
//
// Returns { ok: true, fuelSteps: int } on success or
//         { ok: false, error: "human-readable" } on any of:
//   - unparseable text
//   - intPart < 0 or num < 0 or den < 1
//   - num >= den (a "5/3" frac is invalid)
//   - the fraction's denominator doesn't equal INTERVALS at the wet
//     integer mass (e.g. "1+5/6" with dry=2 → wet integer=3, INTERVALS[3]=4
//     ≠ 6 → reject with a message naming the right denominator).
//   - resulting wet step exceeds 56 (off the strip).
export function parseFuelText(text, dryMass) {
    if (typeof text !== 'string') return { ok: false, error: 'invalid input' };
    const t = text.replace(/\s+/g, '');
    if (t === '') return { ok: true, fuelSteps: 0 };

    // Pattern: optional integer, optional "+num/den" fractional. At least
    // one of the two parts must be present.
    //   "5"        → intPart=5, no frac
    //   "1+5/6"    → intPart=1, num=5, den=6
    //   "5/6"      → intPart=0, num=5, den=6
    //   "+5/6"     → intPart=0, num=5, den=6
    const m = t.match(/^(\d*)(?:\+?(\d+)\/(\d+))?$/);
    if (!m || (!m[1] && !m[2])) {
        return { ok: false, error: 'expected a number like "5" or "1+5/6"' };
    }
    const intPart = m[1] ? parseInt(m[1], 10) : 0;
    const num     = m[2] != null ? parseInt(m[2], 10) : 0;
    const den     = m[3] != null ? parseInt(m[3], 10) : 0;

    if (intPart < 0) return { ok: false, error: 'integer part must be ≥ 0' };
    if (m[2] != null) {
        if (den < 1) return { ok: false, error: 'denominator must be ≥ 1' };
        if (num >= den) return { ok: false, error: `numerator ${num} must be < denominator ${den}` };
        const wetIntMass = (dryMass | 0) + intPart;
        if (wetIntMass < 1 || wetIntMass > 31) {
            return { ok: false, error: `wet integer mass ${wetIntMass} out of strip range` };
        }
        const expectedDen = FUEL_STRIP_INTERVALS[wetIntMass];
        if (den !== expectedDen) {
            return { ok: false, error: `the ${wetIntMass}↔${wetIntMass+1} interval has ${expectedDen} sub-steps, not ${den}` };
        }
    }

    // Fold to fuel-steps: stepsBetween(dryMass, dryMass+intPart) + num
    let fuelSteps = 0;
    for (let i = (dryMass | 0); i < (dryMass | 0) + intPart; i++) {
        if (i < 1 || i > 31) return { ok: false, error: 'fuel value walks off the strip' };
        fuelSteps += FUEL_STRIP_INTERVALS[i];
    }
    fuelSteps += num;
    if (fuelSteps > STRIP_MAX_STEP - massToStripStep(dryMass)) {
        return { ok: false, error: 'wet step exceeds the 56-step strip' };
    }
    return { ok: true, fuelSteps };
}
